File: leetcode3/변지협/165. Compare Version Numbers.py
'''
1. 아이디어 :
    . 을 기준으로 버전 나눠서 비교
2. 시간복잡도 :
    O(n)
3. 자료구조/알고리즘 :
'''

class Solution(object):
    def compareVersion(self, version1, version2):
        """
        :type version1: str
        :type version2: str
        :rtype: int
        """

        # print('111'.find('.'))

        while True:
            if version1 == '' and version2 == '':
                return 0

            v1_idx = version1.find('.')
            v2_idx = version2.find('.')

            # print('version1, version2, v1_idx, v2_idx:', version1, version2, v1_idx, v2_idx)
            # print('version2, len(version2), v2_idx:', version2, len(version2), v2_idx)

            v1_part = ''
            v2_part = ''
            if v1_idx == -1 and len(version1) == 0:
                v1_part = '0'
                version1 = ''
            elif v1_idx == -1 and len(version1) != 0:
                v1_part = version1
                version1 = ''
            else:
                v1_part = version1[:v1_idx]
                version1 = version1[v1_idx+1:]

            if v2_idx == -1 and len(version2) == 0:
                v2_part = '0'
                version2 = ''
            elif v2_idx == -1 and len(version2) != 0:
                v2_part = version2
                version2 = ''
            else:
                v2_part = version2[:v2_idx]
                version2 = version2[v2_idx+1:]

            # print('v1_part, v2_part:', v1_part, v2_part)
            
            if int(v1_part) < int(v2_part):
                return -1
            elif int(v1_part) > int(v2_part):
                return 1


        

File: leetcode3/이진희/165_Compare Version Numbers.java
/* 2차 해결

1. 아이디어 : split("[.]") 로 각 String을 분리한다
            완전탐색을 하며 각 숫자를 비교한다 (이때 삼항연산자로 불필요한 if문 간소화)

2. 시간복잡도 : O(N) (1ms)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public int compareVersion(String version1, String version2) {
        String[] ver1 = version1.split("[.]");
        String[] ver2 = version2.split("[.]");

        int len = Math.max(ver1.length, ver2.length);
        for(int i=0; i<len; i++) {
            
            int v1 = (i<ver1.length) ? Integer.parseInt(ver1[i]) : 0;
            int v2 = (i<ver2.length) ? Integer.parseInt(ver2[i]) : 0;

            if(v1>v2) return 1;
            else if(v1<v2) return -1;
        }
        return 0;
    }
}

/* 1차 해결

1. 아이디어 : split("[.]") 로 각 String을 분리한다
            완전탐색을 하며 각 숫자를 비교한다

2. 시간복잡도 : O(N) (3ms)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public int compareVersion(String version1, String version2) {
        String[] ver1 = version1.split("[.]");
        String[] ver2 = version2.split("[.]");

        int len = Math.max(ver1.length, ver2.length);
        for(int i=0; i<len; i++) {
            if(ver1.length>i && ver2.length>i) {
                int v1 = Integer.parseInt(ver1[i]);
                int v2 = Integer.parseInt(ver2[i]);

                if(v1>v2) return 1;
                else if(v1<v2) return -1;
            }

            else if(ver1.length<=i) {
                if(Integer.parseInt(ver2[i])>0) return -1;
            }
            else if(ver2.length<=i) {
                if(Integer.parseInt(ver1[i])>0) return 1;
            }
        }

        return 0;
    }
}